### PR TITLE
examples: Adds Timer gpout example

### PR DESCRIPTION
**Pull request by saramonteiro**
## Summary

This PR adds one example.
This example uses the timer interrupt to periodically
change the state of a digital output.
The digital output may be a relay, a led or anything else.
This example can be very useful to validate timer drivers
by using a logic analyzer connected to the digital output.
This example, mainly differs from the timer example because it
waits on a sigwaitinfo() instead of using a signal handler.
This approach ensures a deterministic wake-up time when
the signal occurs.

## Impact
N/A

## Testing
Tested and published a video in NuttX Brasil channel running this example with a LA:
https://youtu.be/v4bSGF3b3v4

### CODE REVIEW

File: examples/timer_gpout/Make.defs
Comment (patacongo) [inline examples/timer_gpout/Make.defs:35]: Shouldn't this be an ASF Apache 2.0 header?

File: examples/timer_gpout/Makefile
Comment (patacongo) [inline examples/timer_gpout/Makefile:35]: Shouldn't this be an ASF Apache 2.0 header?

File: examples/timer_gpout/timer_gpout_main.c
Comment (patacongo) [inline examples/timer_gpout/timer_gpout_main.c:53]: The NuttX coding standard requires that all variables defined at global scope must begin with g_.  Putting _g at the end violates the coding standard.  These need to be changed to g_devtim, g_devgpout, and g_timer_gpout_daemon_started.

File: examples/timer_gpout/timer_gpout_main.c
Comment (patacongo) [inline examples/timer_gpout/timer_gpout_main.c:219]:
```suggestion
      state = !state;
```
The tilde is a bit operation of a boolean value.  It will toggle from 0 (false) and 0xffffffff (invalid boolean value).  The exclamation point is the correct operator for a boolean and will toggle from 0 (false) to 1 (true).

File: examples/timer_gpout/Make.defs
Comment (saramonteiro) [reply examples/timer_gpout/Make.defs:19]: Yes.
Solved.
Thanks.

File: examples/timer_gpout/Makefile
Comment (saramonteiro) [reply examples/timer_gpout/Makefile:20]: Yes.
Solved.
Thanks.

File: examples/timer_gpout/timer_gpout_main.c
Comment (saramonteiro) [reply examples/timer_gpout/timer_gpout_main.c]: I inverted. Fixed.

File: examples/timer_gpout/timer_gpout_main.c
Comment (saramonteiro) [reply examples/timer_gpout/timer_gpout_main.c]: Ok.
Fixed.

**Comment by saramonteiro** [comment]
@patacongo I fixed the found issues. PTAL.
Note: I'll squash the commits once it's approved.
Do not merge without it.

**Review by patacongo — APPROVED** [review]
Thanks!

**Comment by patacongo** [comment]
A question.  This is not essential to the PR, but is it really necessary to limit the number of daemons waiting for the signal?  Is that a limitation of the timer? What would happen if you removed g_timer_gpout_daemon_started so that there could be multiple active timers?

**Comment by saramonteiro** [comment]
> A question. This is not essential to the PR, but is it really necessary to limit the number of daemons waiting for the signal? Is that a limitation of the timer? What would happen if you removed g_timer_gpout_daemon_started so that there could be multiple active timers?

This a good question. Although I haven't tested, I infer that we can't have multiple deamons with one timer. Because each time one daemon was launched it would rewrite the `timer_notify_s` structure which is used within the timer callback to send a signal for a specific process id (task).  As a consequence, I believe we should see the priors daemons starve. To achieve this purpose, I believe we would have to create a Task Group and set up signaling for Multi-threaded Task Group. (What's is actually a good idea for a next example...).
But I see no reasons we shouldn't have multiple daemons (each one associated with a specific timer).
I'll perform both test and I come back here to tell you.

**Comment by patacongo** [comment]
I think I should go ahead a merge this version.  It would be a good extension to testing to prove that there is no issue with multiple concurrent timers.  It would be a bug in the driver if it could not support more than one client task.

But, I think you can extend this testing in any way that you wish in the future.  Let's merge this well verified PR as it is now.

**Comment by patacongo** [comment]
The driver really should handle multiple clients like all other drivers.  That should be a requirement of all drivers so I would personally consider that limitation a bug.

For a single timer, this would be handled by keeping a list of relative timer delays.  This is done, for example, for the wdog dog timers.  See https://github.com/apache/incubator-nuttx/blob/master/sched/wdog/wd_start.c#L226

**Comment by patacongo** [comment]
> The driver really should handle multiple clients like most other drivers. That should be a requirement of all drivers, if possible, so I would personally consider that limitation a bug in this case.

Ha ha... looks like I wrote the driver/timer/timer.c driver way back in 2012.  So I guess if it is a bug, it is my bug.

**Comment by saramonteiro** [comment]
@patacongo I performed some tests and here are the tests with the results:

**1st Test:**
Simply removed the part that checks for `g_timer_gpout_daemon_started`.
Result:
Simply doesn't work, because when it calls the timer start ioctl, it returns EBUSY.

**2nd Test**
Kept the changes from the 1st test, but after opening timer driver, I called the status ioctl to use the status flags and check the TCFLAGS_ACTIVE. Then I checked If the timer was running, in case it was, I set it to perform only the notification ioctl and not to perform settimeout and start.
Result:
It works, but as I expected, `timer_notify_s` structure was subscribed and the previous daemon starved. Take a look:
image
Conclusion for this: For one timer, we should have only one daemon.

**3rd Test**
I kept _only_ the changes from the 1st test and ran multiple daemons. Each one using one timer and a specific gpout. Ex:
`timer_gpout -t /dev/timer0 -g /dev/gpout0`
`timer_gpout -t /dev/timer1 -g /dev/gpout1`
Result:
It works perfectly: 
image

**Pull request merged**
